### ISSUE TITLE
Confused about usertypes and function parameters

### ISSUE THREAD

**Issue by mries92** [ThePhD/sol2#1236]
Sorry to post this as an issue, but I have been struggling with this for a while and can't seem to figure it out. I have been able to register C++ functions that take primitive types as parameters, but when I try and register a function that takes a custom struct, it never seems to hit the function. I expect I am probably doing something wrong with registering my UserType. Here is the process I am taking.

Registering the UserType
```c++
sol::table renderer_table = lua.create_named_table("renderer");
renderer_table.set_function("add_drawable", &Game::LuaAddDrawable, this);
sol::usertype<LuaDrawableDetails> drawable_usertype = lua.new_usertype<LuaDrawableDetails>("drawable_details", sol::constructors<LuaDrawableDetails(), LuaDrawableDetails(std::string, std::string)>());
drawable_usertype["type"] = &LuaDrawableDetails::type;
drawable_usertype["sprite_id"] = &LuaDrawableDetails::sprite_id;
```

The type
```c++
struct LuaDrawableDetails
{
    LuaDrawableDetails() {};
    LuaDrawableDetails(std::string type, std::string sprite_id) { this->type = type; this->sprite_id = sprite_id; };

    std::string type;
    std::string sprite_id;
};
```

The c++ method where I am using this struct
```c++
void Game::LuaAddDrawable(LuaDrawableDetails details)
```

And finally the actual Lua call (this call goes through if I register the function as taking no parameters, so my registration is at least correct)
```lua
renderer.add_drawable({type = "sprite", sprite_id = "player"})
```

I am probably doing something stupid. Any help would be appreciated, and thanks for the awesome library!

**Comment by wzhengsen** [comment]
```c++
#include "sol/sol.hpp"
#include <string>
#include <iostream>

namespace Game {
    struct LuaDrawableDetails {
        LuaDrawableDetails() {};
        LuaDrawableDetails(std::string type, std::string sprite_id) { this->type = type; this->sprite_id = sprite_id; };

        std::string type;
        std::string sprite_id;
    };

    void LuaAddDrawable(LuaDrawableDetails details) {
        std::cout << "type:" << details.type << std::endl << "sprite_id:" << details.sprite_id << std::endl;
    }
};

template <typename Handler>
bool sol_lua_check(sol::types<Game::LuaDrawableDetails>, lua_State* L, int index, Handler&& handler, sol::stack::record& tracking) {
    auto success = sol::stack::check<sol::lua_table>(L, index, std::forward<Handler>(handler));
    if (success) {
        success = LUA_TSTRING == lua_getfield(L, index, "type");
        lua_pop(L, 1);
    }
    if (success) {
        success = LUA_TSTRING == lua_getfield(L, index, "sprite_id");
        lua_pop(L, 1);
    }
    tracking.use(1);
    return success;
}
Game::LuaDrawableDetails sol_lua_get(sol::types<Game::LuaDrawableDetails>, lua_State* L, int idx, sol::stack::record& tracking) {
    sol::table t = sol::table(L, idx);
    std::string _type = t["type"];
    std::string _sprite_id = t["sprite_id"];
    tracking.use(1);
    return { _type,_sprite_id };
}

int main() {
    sol::state lua = {};
    lua.open_libraries();
    lua["LuaAddDrawable"] = &Game::LuaAddDrawable;

    //output:
    //type:sprite
    //sprite_id:player
    lua.script("LuaAddDrawable({type = 'sprite', sprite_id = 'player'});");

    //error
    lua.script("LuaAddDrawable({type1 = 'sprite', sprite_id = 'player'});");
    return 0;
}
```

**Comment by wzhengsen** [comment]
Or you may not register the type for LuaDrawableDetails:
```c++
#include "sol/sol.hpp"
#include <string>
#include <iostream>

struct LuaDrawableDetails {
    LuaDrawableDetails() {};
    LuaDrawableDetails(std::string type, std::string sprite_id) { this->type = type; this->sprite_id = sprite_id; };

    std::string type;
    std::string sprite_id;
};

namespace Game {
    void LuaAddDrawable(LuaDrawableDetails details) {
        std::cout << "type:" << details.type << std::endl << "sprite_id:" << details.sprite_id << std::endl;
    }
};

template <typename Handler>
bool sol_lua_check(sol::types<LuaDrawableDetails>, lua_State* L, int index, Handler&& handler, sol::stack::record& tracking) {
    auto success = sol::stack::check<sol::lua_table>(L, index, std::forward<Handler>(handler));
    if (success) {
        success = LUA_TSTRING == lua_getfield(L, index, "type");
        lua_pop(L, 1);
    }
    if (success) {
        success = LUA_TSTRING == lua_getfield(L, index, "sprite_id");
        lua_pop(L, 1);
    }
    tracking.use(1);
    return success;
}
LuaDrawableDetails sol_lua_get(sol::types<LuaDrawableDetails>, lua_State* L, int idx, sol::stack::record& tracking) {
    sol::table t = sol::table(L, idx);
    std::string _type = t["type"];
    std::string _sprite_id = t["sprite_id"];
    tracking.use(1);
    return { _type,_sprite_id };
}

int main() {
    sol::state lua = {};
    lua.open_libraries();
    lua["LuaAddDrawable"] = &Game::LuaAddDrawable;

    //output:
    //type:sprite
    //sprite_id:player
    lua.script("LuaAddDrawable({type = 'sprite', sprite_id = 'player'});");

    //error
    lua.script("LuaAddDrawable({type1 = 'sprite', sprite_id = 'player'});");
    return 0;
}
```

**Comment by mries92** [comment]
Dude..... you are my hero. I was on [this page](https://sol2.readthedocs.io/en/latest/tutorial/customization.html) skimming it for a bit, and couldn't quite tell the difference between what it was doing and UserTypes. After about 30 minutes of fiddling I got both of your solutions working. I prefer the second method as it feels more "Luaish", and fits in better with the API I have exposed from the public game files.

Thank you so much! Time to get back to working on the game 😄
